src/handlers/permissions-revoke.ts

```diff
@@ -0,0 +1,115 @@
+    let permissionsRevoke: PermissionsRevoke;
+    try {
+      permissionsRevoke = await PermissionsRevoke.parse(message);
+    } catch (e) {
+      return messageReplyFromError(e, 400);
+    }
+
+
+    // Authentication
+    try {
+      await authenticate(message.authorization, this.didResolver);
+    } catch (e) {
+      return messageReplyFromError(e, 401);
+    }
+
+    // Authorization
+    const permissionsGrantId = message.descriptor.permissionsGrantId;
+    const permissionsGrantMessage = await this.messageStore.get(
+      tenant,
+      permissionsGrantId
+    ) as PermissionsGrantMessage | undefined;
+
+    if (permissionsGrantMessage === undefined) {
+      return {
+        status: {
+          code   : 400,
+          detail : `Could not find PermissionsGrant with CID ${permissionsGrantId}`
+        }
+      };
+    }
+
+    try {
+      await permissionsRevoke.authorize(permissionsGrantMessage);
+    } catch (e) {
+      return messageReplyFromError(e, 401);
+    }
+
+    // Revoke must have `dateCreated` after that of the grant
+    if (message.descriptor.dateCreated < permissionsGrantMessage.descriptor.dateCreated) {
+      return {
+        status: { code: 400, detail: 'PermissionsRevoke has earlier date than associated PermissionsGrant' }
+      };
+    }
+
+    // query for previous revocations of this grant
+    const query = {
+      interface : DwnInterfaceName.Permissions,
+      method    : DwnMethodName.Revoke,
+      permissionsGrantId,
+    };
+    const existingRevokesForGrant: PermissionsRevokeMessage[] = await this.messageStore.query(tenant, query) as PermissionsRevokeMessage[];
+
+    // Conflict 409 if the grant already has an older revoke
+    const oldestExistingRevoke = PermissionsRevoke.getOldestRevoke(existingRevokesForGrant);
+    if (oldestExistingRevoke !== undefined && oldestExistingRevoke.descriptor.dateCreated < message.descriptor.dateCreated) {
+      return {
+        status: { code: 409, detail: 'Conflict' }
+      };
+    }
+
+    // Store incoming PermissionsRevoke
+    const indexes: { [key: string]: string } = {
+      interface          : DwnInterfaceName.Permissions,
+      method             : DwnMethodName.Revoke,
+      permissionsGrantId : message.descriptor.permissionsGrantId,
+    };
+    await this.messageStore.put(tenant, message, indexes);
+    await this.eventLog.append(tenant, await Message.getCid(message));
+
+    // Delete existing revokes which are all older than the incoming message
+    const removedRevokeCids: string[] = [];
+    for (const existingRevoke of existingRevokesForGrant) {
+      const existingRevokeCid = await Message.getCid(existingRevoke);
+      await this.messageStore.delete(tenant, existingRevokeCid);
+      removedRevokeCids.push(existingRevokeCid);
+    }
+    await this.eventLog.deleteEventsByCid(tenant, removedRevokeCids);
+
+    // Delete grant-authorized messages with timestamp after revocation
```

Comment:
Related to https://github.com/TBD54566975/dwn-sdk-js/pull/417/files?diff=unified&w=0#r1245751682

To make the scenario a bit more complex, consider the following sequence of events:
1. Alice issues a `PermissionsGrant` to Bob, and all of her DWNs have processed this grant.
2. Alice sends a `PermissionsRevoke` to her local DWN
3. Bob sends a message to Alice's remote DWN using his `PermissionsGrant`
4. Alice's remote DWN syncs with her local DWN, pulling the `PermissionsRevoke`.

Since the revocation was created before Bob's message, Bob's message should not actually have been authorized. So, Bob's message is deleted from Alice's DWN.

Reply:
Are you implying that processing Bob's request at the remote DWN is delayed until it syncs with Alice's local DWN?  What if that sync is delayed?

The important thing to note is that there is an unavoidable race between Alice's revoke and Bob's message because their actions are concurrent.  After all, Bob could have sent his message just a little bit before Alice sent her revoke or just a little bit after.  That means it's important to carefully define what "before" means in this context.

Reply:
No, Bob's request is not delayed. Alice's remote DWN processes Bob's request. Some time later Alice's DWNs sync, and her remote DWN receives the `PermissionsRevoke`, which has an earlier `messageTimestamp` than Bob's message. Even though Alice's remote processed Bob's message first, the `messageTimestamp` takes precedence, and Bob's message is deleted.

Reply:
If Bob's message is processed, then any action it specifies is processed as if no revoke action had taken place.  If you later delete Bob's message, does that undo any action taken when processing Bob's message?